src/fs/fuse.rs

```diff
@@ -38,1190 +38,557 @@ const U64_SIZE: usize = ::core::mem::size_of::<u64>();
-
-#[repr(C)]
-#[derive(Debug)]
-struct fuse_dirent {
-	pub d_ino: u64,
-	pub d_off: u64,
-	pub d_namelen: u32,
-	pub d_type: u32,
-	pub d_name: [u8; 0],
-}
-
 pub(crate) trait FuseInterface {
-	fn send_command<S, T>(&mut self, cmd: &Cmd<S>, rsp: &mut Rsp<T>)
-	where
-		S: FuseIn + core::fmt::Debug,
-		T: FuseOut + core::fmt::Debug;
+	fn send_command<const CODE: u32>(
+		&mut self,
+		cmd: &<Op<CODE> as OpTrait>::Cmd,
+		rsp: &mut <Op<CODE> as OpTrait>::Rsp,
+	) where
+		Op<CODE>: OpTrait;
 
 	fn get_mount_point(&self) -> String;
 }
 
-#[repr(C)]
-#[derive(Debug, Default)]
-struct fuse_in_header {
-	pub len: u32,
-	pub opcode: u32,
-	pub unique: u64,
-	pub nodeid: u64,
-	pub uid: u32,
-	pub gid: u32,
-	pub pid: u32,
-	pub padding: u32,
-}
-
-#[repr(C)]
-#[derive(Default, Debug)]
-struct fuse_out_header {
-	pub len: u32,
-	pub error: i32,
-	pub unique: u64,
-}
+/// Marker trait, which signals that a struct is a valid Fuse command.
+/// Struct has to be repr(C)!
+pub(crate) unsafe trait FuseIn {}
+/// Marker trait, which signals that a struct is a valid Fuse response.
+/// Struct has to be repr(C)!
+pub(crate) unsafe trait FuseOut {}
 
-#[repr(C)]
-#[derive(Debug, Default)]
-struct fuse_init_in {
-	pub major: u32,
-	pub minor: u32,
-	pub max_readahead: u32,
-	pub flags: u32,
+unsafe impl FuseIn for fuse_abi::InitIn {}
+unsafe impl FuseOut for fuse_abi::InitOut {}
+unsafe impl FuseIn for fuse_abi::ReadIn {}
+unsafe impl FuseIn for fuse_abi::WriteIn {}
+unsafe impl FuseOut for fuse_abi::WriteOut {}
+unsafe impl FuseOut for fuse_abi::ReadOut {}
+unsafe impl FuseIn for fuse_abi::LookupIn {}
+unsafe impl FuseIn for fuse_abi::ReadlinkIn {}
+unsafe impl FuseOut for fuse_abi::ReadlinkOut {}
+unsafe impl FuseOut for fuse_abi::AttrOut {}
+unsafe impl FuseOut for fuse_abi::EntryOut {}
+unsafe impl FuseIn for fuse_abi::CreateIn {}
+unsafe impl FuseOut for fuse_abi::CreateOut {}
+unsafe impl FuseIn for fuse_abi::OpenIn {}
+unsafe impl FuseOut for fuse_abi::OpenOut {}
+unsafe impl FuseIn for fuse_abi::ReleaseIn {}
+unsafe impl FuseOut for fuse_abi::ReleaseOut {}
+unsafe impl FuseIn for fuse_abi::RmdirIn {}
+unsafe impl FuseOut for fuse_abi::RmdirOut {}
+unsafe impl FuseIn for fuse_abi::MkdirIn {}
+unsafe impl FuseIn for fuse_abi::UnlinkIn {}
+unsafe impl FuseOut for fuse_abi::UnlinkOut {}
+unsafe impl FuseIn for fuse_abi::LseekIn {}
+unsafe impl FuseOut for fuse_abi::LseekOut {}
+unsafe impl FuseIn for fuse_abi::PollIn {}
+unsafe impl FuseOut for fuse_abi::PollOut {}
+
+pub(crate) trait OpTrait {
+	type InStruct: FuseIn + core::fmt::Debug;
+	type InPayload: ?Sized;
+	type OutStruct: FuseOut + core::fmt::Debug;
+	type OutPayload: ?Sized;
+
+	type Cmd: ?Sized + AsSliceU8 = ReqPart<fuse_abi::InHeader, Self::InStruct, Self::InPayload>;
+	type Rsp: ?Sized + AsSliceU8 =
+		ReqPart<fuse_abi::OutHeader, MaybeUninit<Self::OutStruct>, Self::OutPayload>;
 }
 
-unsafe impl FuseIn for fuse_init_in {}
+pub(crate) struct Op<const CODE: u32>;
```

Comment:
Is there any benefit of having this be a const generic instead of having distinct types that implement `OpTrait`? I found that `Op<{ fuse_abi::Opcode::Init as u32 }>` reads much more difficult than, just `ops::Init`. Similar for where clauses where `cmd: &T::Cmd where T: OpTrait` would be much easier to read than `cmd: &<fuse::Op<CODE> as fuse::OpTrait>::Cmd where fuse::Op<CODE>: fuse::OpTrait`.

Reply:
You are right. I started writing the code thinking that C-style enums can be const generic parameters and once it didn't work as I hoped I worked around the limitation instead of changing it to empty structs.

Reply:
Solved with 695a46766e5e2947b3b71af8d65f6740035e27ee.